hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientResult.java

```diff
@@ -0,0 +1,181 @@
+        this.service = service;
+        this.connection = connection;
+        this.queryId = queryId;
+        this.rowMetadata = rowMetadata;
+        this.cursorBufferSize = cursorBufferSize;
+
+        iterator.onNextPage(page);
+    }
+
+    @Nonnull
+    @Override
+    public SqlRowMetadata getRowMetadata() {
+        return rowMetadata;
+    }
+
+    @Override
+    @Nonnull
+    public Iterator<SqlRow> iterator() {
+        if (!iteratorAccessed) {
+            iteratorAccessed = true;
+
+            return iterator;
+        } else {
+            throw new IllegalStateException("Iterator could be requested only once");
+        }
+    }
+
+    @Override
+    public void close() {
+        try {
+            if (!closed) {
+                if (iterator.last) {
+                    // The last page accessed, so the remote cursor is already closed. No-op.
+                    return;
+                }
+
+                service.close(connection, queryId);
+            }
+        } finally {
+            closed = true;
+        }
+    }
+
+    public int getCursorBufferSize() {
+        return cursorBufferSize;
+    }
+
+    public void setCursorBufferSize(int cursorBufferSize) {
+        assert cursorBufferSize >= 0;
+
+        this.cursorBufferSize = cursorBufferSize;
+    }
+
+    private void fetchNextPage(ClientIterator iterator) {
+        SqlPage page = service.fetch(connection, queryId, cursorBufferSize);
+
+        iterator.onNextPage(page);
+    }
+
+    private List<Row> convertPageRows(List<SqlPageRow> serializedRows) {
+        List<Row> rows = new ArrayList<>(serializedRows.size());
+
+        for (SqlPageRow serializedRow : serializedRows) {
+            List<Data> serializedValues = serializedRow.getValues();
+            Object[] values = new Object[serializedValues.size()];
+
+            for (int i = 0; i < serializedValues.size(); i++) {
+                values[i] = service.deserializeRowValue(serializedValues.get(i));
+            }
+
+            rows.add(new HeapRow(values));
+        }
+
+        return rows;
+    }
+
+    /**
+     * Implementation of lazy iterator, which fetches results as needed.
+     */
+    private class ClientIterator implements Iterator<SqlRow> {
+
+        private List<Row> currentRows;
+        private int currentPosition;
+        private boolean last;
+
+        @Override
+        public boolean hasNext() {
+            if (closed) {
+                throw service.rethrow(QueryException.cancelledByUser());
+            }
+
+            while (currentPosition == currentRows.size()) {
+                // Reached end of the page. Try fetching the next one if possible.
+                if (!last) {
+                    fetchNextPage(this);
+                } else {
+                    // No more pages expected, so return false.
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        @Override
+        public SqlRow next() {
```

Comment:
Shall we add if(closed) check here as well?

Reply:
This method first invokes `hasNext` that does this check.